source/array_list_with_map.c

```diff
@@ -0,0 +1,127 @@
+    }
+    size_t current_length = aws_array_list_length(&list_with_map->list);
+    if (aws_array_list_push_back(&list_with_map->list, element)) {
+        goto list_push_error;
+    }
+    if (aws_hash_table_put(&list_with_map->map, element, (void *)current_length, NULL)) {
+        goto error;
+    }
+    return AWS_OP_SUCCESS;
+error:
+    aws_array_list_pop_back(&list_with_map->list);
+list_push_error:
+    return AWS_OP_ERR;
+}
+
+int aws_array_list_with_map_remove(struct aws_array_list_with_map *list_with_map, const void *element) {
+    AWS_PRECONDITION(list_with_map);
+    AWS_PRECONDITION(element);
+    size_t current_length = aws_array_list_length(&list_with_map->list);
+    if (current_length == 0) {
+        /* Nothing to remove */
+        return AWS_OP_SUCCESS;
+    }
+    /* find the index of element first */
+
+    struct aws_hash_element *find = NULL;
+    /* find and remove the elment from table */
+    if (aws_hash_table_find(&list_with_map->map, element, &find)) {
+        return AWS_OP_ERR;
+    }
+    if (!find) {
+        /* It's removed already */
+        return AWS_OP_SUCCESS;
+    }
+
+    size_t index_to_remove = (size_t)find->value;
+    if (aws_hash_table_remove(&list_with_map->map, element, NULL, NULL)) {
+        return AWS_OP_ERR;
+    }
+    /* Nothing else can fail after here. */
+
+    if (index_to_remove != current_length - 1) {
+        /* It's not the last element, we need to swap it with the end of the list and remove the last element */
+        void *last_element = aws_mem_acquire(list_with_map->list.alloc, list_with_map->list.item_size);
+        AWS_FATAL_ASSERT(aws_array_list_back(&list_with_map->list, last_element) == AWS_OP_SUCCESS);
+        /* Update the last element index in the table */
+        struct aws_hash_element *element_to_update = NULL;
+        AWS_FATAL_ASSERT(aws_hash_table_find(&list_with_map->map, last_element, &element_to_update) == AWS_OP_SUCCESS);
+        AWS_FATAL_ASSERT(element_to_update != NULL);
+        element_to_update->value = (void *)index_to_remove;
+        /* Swap the last element with the element to remove in the list */
+        aws_array_list_swap(&list_with_map->list, index_to_remove, current_length - 1);
+        aws_mem_release(list_with_map->list.alloc, last_element);
+    }
+    /* Remove the current last element from the list */
+    AWS_FATAL_ASSERT(aws_array_list_pop_back(&list_with_map->list) == AWS_OP_SUCCESS);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_array_list_with_map_get_random(struct aws_array_list_with_map *list_with_map, void *out) {
+    AWS_PRECONDITION(list_with_map);
+    size_t length = aws_array_list_length(&list_with_map->list);
+    if (length == 0) {
+        return aws_raise_error(AWS_ERROR_LIST_EMPTY);
+    }
+
+    /* use the best of two algorithm to select the connection with the lowest load. */
+    uint64_t random_64_bit_num = 0;
+    aws_device_random_u64(&random_64_bit_num);
+    size_t index = random_64_bit_num % length;
```

Comment:
nit: the result of mod will be skewed and you would not get ideal random distribution. consider something like http://c-faq.com/lib/randrange.html

Reply:
I looked deeper into aws_device_random and it claims to provide random numbers usable for crypto purposes. In that case, i dont think mod vs div matters that much.
Concern with mod is that a lot of default rand primitives generate numbers that will have poor randomness in the lower bits and taking mod of those will have non-uniform distribution, while crypto safe rand should have even distribution through the bits. So both approaches will have similar distributions. 
Note: unless length is multiple of u64 max all approaches will have some "mod skew", but it will be tiny (something like a couple of elements at the start of the list having higher probability by 1 / 2^~62

Reply:
We should do fuzz tests and know for sure. Get 1,000,000 random indices, and assert that the distributions is within 3 or 4 standard deviations of normal. Note each time we run CI we would be running this test ~20 times so we need to tolerate a decent amount of variance.

Reply:
We already have tests for random number gen https://github.com/awslabs/aws-c-common/blob/057746b2e094f4b7a31743d8ba5a9fd0155f69f3/tests/device_random_test.c#L34. Looks like they bucket generated numbers and check that buckets are uniform. Michael implemented those tests and he chose div approach instead of mod, might be worth checking with him if one is better choice than the other